geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/GroupBy.scala

```diff
@@ -0,0 +1,113 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.utils.stats
+
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+case class GroupBy[T](attribute: Int, exampleStat: Stat)(implicit ct: ClassTag[T]) extends Stat {
+  override type S = GroupBy[T]
+
+  val groupedStats: mutable.HashMap[T, Stat] = mutable.HashMap[T, Stat]()
+
+  /**
+    * Compute statistics based upon the given simple feature.
+    * This method will be called for every SimpleFeature a query returns.
+    *
+    * @param sf feature to evaluate
+    */
+  override def observe(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key) match {
+      case Some(groupedStat) => groupedStat.observe(sf)
+      case None              => val newStat = exampleStat.newcopy
+        newStat.observe(sf)
+        groupedStats.update(key, newStat)
+    }
+  }
+
+  /**
+    * Tries to remove the given simple feature from the compiled statistics.
+    * Note: may not be possible to un-observe a feature, in which case this method will
+    * have no effect.
+    *
+    * @param sf feature to un-evaluate
+    */
+  override def unobserve(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key) match { case Some(groupedStat) => groupedStat.unobserve(sf) }
+  }
+
+  /**
+    * Add another stat to this stat. Avoids allocating another object.
+    *
+    * @param other the other stat to add
+    */
+  override def +=(other: GroupBy[T]): Unit = {
+    other.groupedStats.map { case (key, stat) =>
+      groupedStats.get(key) match {
+        case Some(groupedStat) => groupedStat += stat
+        case None              => groupedStats.put(key, stat)
+      }
+    }
+  }
+
+  /**
+    * Combine two stats into a new stat
+    *
+    * @param other the other stat to add
+    */
+  override def +(other: GroupBy[T]): GroupBy[T] = {
+    val newGB = new GroupBy[T](attribute, exampleStat.newcopy)
+    newGB += this
+    newGB += other
+    newGB
+  }
+
+  /**
+    * Returns a json representation of the stat
+    *
+    * @return stat as a json string
+    */
+  override def toJson: String = {
+    // TODO: Provide summary of all items in all groups together. e.g. GroupBy("cat",Count()) should have a total count as well as a per category count.
+    groupedStats.map{ case (key, stat) => "{ \"" + key + "\" : " + stat.toJson + "}" }.mkString("[",",","]")
```

Comment:
Since this is a HashMap will the key order be deterministic?  If not, is this ok?  I would think this would be desirable to the end-user.

```
ListMap(hm.toSeq.sortBy(_._1):_*).map { ...
```